### DISCUSSION TITLE
postgresql function called from client doesn't generate expected result as in sql editor

### DISCUSSION THREAD

**Discussion by JinNgVN** [supabase/supabase#20265] [Questions]
Hi guys, 
So i create a fucntion called get_random_models() like this 
```
create or replace function get_random_models()
returns setof models
language sql
as $$
 select * from models order by random() limit 6
$$
```

When i use it in the sql editor like this: 
`select * from get_random_models()`
the result is correct. 
However, when i try to call it from client side:
`const { data, error } = await supabase.rpc("get_random_models").select("*");`
 The result is the same every time the request is made. I also tried 
`const { data, error } = await supabase.rpc("get_random_models");` 
but it only return 1 object, not even an array while my table have 6 rows. Do you know what is wrong with supabase client?

Thank you in advance.

**Comment by GaryAustin1** [comment]
The main difference in the client versus the SQL editor (with the default Postgres users) is that the client obeys RLS for your function.

Also though you comment about getting the same result each time.  Are you using next.js?  It caches URL's in many cases and leads to things like that unless you disable the caching.  The Supabase clients and server do no caching.

**Comment by alt-glitch** [comment]
Hey @JinNgVN could you share if you have any RLS policies enabled?
The SQL editor bypasses RLS policies and the client might not have access to certain rows if it's `anon` or not authenticated.

Is the result of `const { data, error } = await supabase.rpc("get_random_models").select("*");`  just a single same object always?
What is the result when you reinitialise the supabase client ie, does it stay the same?

I ask because I made the same function and was able to get random objects on every call.

**Comment by JinNgVN** [comment]
Hi @alt-glitch and @GaryAustin1,
Sorry for late reply.
So I only have one RLS enabled for that table which is "enable read access to everyone".
So i modified my code a bit but the result is still the same.
I defined my function: get_random_models() as following:
```
create or replace function get_random_models()
returns setof json
language sql
as $$
  SELECT json_build_object(
    'model_id', m.model_id,
    'name', m.name,
   'description', m.description,
   'main_picture_url', m.main_picture_url
  )
  FROM models m
  ORDER BY random()
  LIMIT 6
$$
```
When i run it in sql editor, i got the correct result every time i call it using this command:
`select get_random_models() `
It generates a list of object and it is random everytime.
However, when i call it using supabase client like this:
```
 const { data, error } = await supabase.rpc("get_random_models")
  console.log(data)
```
Then i only get only one object, not even an array of one object.
But if i modify my code to this:
`
const { data, error } = await supabase.rpc("get_random_models").select("*");
`
Then i get the an array of 6 objects but they are the same everytime i reset the page, crtl + s, reload, ...
I dont know what i am missing!
Thank you everyone for helping me

**Reply by JinNgVN** [reply]
I thought if i want to call a function, then `supabase.rpc("get_random_models")` should be enough and do not need to add .select(" * ") like this.

**Reply by JinNgVN** [reply]
i have created some other functions like get_lastest_videos (this one is not with "order by random()"). All i need to do is to use:
`supabase.rpc("get_lastest_videos")`
Then it returns a list of videos without adding the .select statement
